### ISSUE TITLE
Aspect ratio not being honored for video player sidecar poster images

### ISSUE THREAD

**Issue by JesseHC** [zenphoto/zenphoto#1381]
Using the native html video player causes widescreen sidecar poster images to be cropped to the 4:3 aspect ratio. By default, sidecar poster images are being cropped to 480x360. I don’t see an option to change this under Image, Theme, or Plugin options. The class-video plugin doesn’t give you any options other than to set a watermark or not.

If you use the mediaelementjs_player (v1.1.1), you can set the video poster size, but you only get to pick one size. The poster image will then be stretched if it is not in that aspect ratio.

Ideally, sidecar poster images for videos would be treated just like other images when set to not be cropped, and scaled to the video player correctly. That way, if your site has both 16:9 and 4:3 videos on it, they will all display correctly.

**Comment by JesseHC** [comment]
It looks like if we can get Zenphoto to stop cropping and shrinking the video poster images, the answer is to add this to the theme css to maintain the poster image aspect ratio:

```
video {
  object-fit: contain;
}
```

More info-
https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit

Idea from here-
https://stackoverflow.com/a/31044171

_I tested it in Firefox 111.0, Chrome 111.0.5563.64, and Safari 14.1.2 (14611.3.10.1.7)._

**Comment by acrylian** [comment]
Yes, but IMHO it is theme domain to do this. Does this really work on the poster image attribute? Otherwise it might affect the video itself perhaps?

**Comment by acrylian** [comment]
I have added some things and some options regarding audio and video posters.

**Comment by JesseHC** [comment]
Thanks for the effort, but it’s not working.

First, I don’t think poster images should be cropped, ever. They should be set to contain in the video player so they are always at the right aspect ratio. They could be shrunk to the video player size, if set, but again that is not an option, and if they are going to be playing the video, you might as well keep the poster at the same resolution.

As it is right now if you have a 640x480 4:3 SD video with a poster, and have the poster size set in class-video to 16:9 720p HD, so 1280x720, it will take that 640x480 poster image, stretch the width to 1280, and for some reason, crop it to half height at 360.

I can later try to whip up some standard video resolution files with poster images if it will help you test.

**Comment by JesseHC** [comment]
To help test different video sizes, aspect ratios, and sidecar images, I created this basic pack of sample videos.

It includes the following standard sizes in MP4 format with sidecar PNG images:

| Type 	| Name 	| Aspect Ratio 	| Pixel Size |
| --- | --- | --- | --- |
| SD (Standard Definition) 	| 480p 		| 4:3 	| 640 x 480 |
| HD (High Definition) 		| 720p 		| 16:9 	| 1280 x 720 |
| Full HD (FHD) 		| 1080p 	| 16:9 	| 1920 x 1080 |
| 4K video or Ultra HD (UHD) 	| 4K or 2160p 	| 1:1.9 | 3840 x 2160 |
| Instagram			| Square	| 1:1 	| 1080 x 1080 |
| Facebook or TikTok		| Portrait 	| 9:16 	| 1080 x 1920 |

While there are many more resolution formats and aspects ratios out there, this will give you a really good base to test from.

[sample_video_sizes.zip](https://github.com/zenphoto/zenphoto/files/11131732/sample_video_sizes.zip)

![Screenshot of sample_video_sizes](https://user-images.githubusercontent.com/82089671/229344999-c2d9bee2-b0da-4630-86ad-fd4580efc267.png)

In a default Zenphoto Basic theme album with cropping turned off-

![Basic - with crop off](https://user-images.githubusercontent.com/82089671/229345155-505382ee-5506-4cee-a890-9a59b8e8d9ad.png)

**Comment by JesseHC** [comment]
While looking into it, I came across this old bug report I filed years ago under my work account regarding sidecar thumbs being cropped-

https://github.com/zenphoto/zenphoto/issues/1132

You can see there that sidecar thumbs were being cropped, and it was fixed. The different video plugins didn’t really need it, since they offered their own poster sizes and poster image handling (though there are still issues with them), but the native poster handling was never fixed to remove crops.

Looking through the cache, besides Zenphoto with the sidecar thumbs, html_meta_tags is also properly scaling the Open graph images.

So what it looks like is needed for the video poster images to work properly is to turn off the crop, and set the MaxSpaceContainer so they are properly scaled.

I would also recommend setting the default poster size for class-video to 1280x720, since that is what YouTube uses for their thumbnails, is the smallest HD size, and is widescreen.

https://support.google.com/youtube/answer/72431?hl=en#zippy=%2Cimage-size-resolution

You can even test this by registering a 1280x720 size with MaxSpace set in cacheManager and running it on just the sidecar images. Then replace the video posters in your browser’s inspector, and you will see that they work fine.


Video posters being created in the current updated class-video-

![Basic - Portrait](https://user-images.githubusercontent.com/82089671/229345680-2ccd19b9-ff1d-4fd2-80a0-24d80035ae32.png)

![Basic - SD](https://user-images.githubusercontent.com/82089671/229345688-3f47b1e0-56a2-4c89-a9d9-2cce1c3e6360.png)

![Basic - Square](https://user-images.githubusercontent.com/82089671/229345694-ee7900b2-b435-4223-9b43-c58f225bab4e.png)


Video posters created in cacheManager at 1280x720 with no crop and MaxSpace set, and then replaced in the browser-

![Portrait replaced](https://user-images.githubusercontent.com/82089671/229345702-23da71d9-e462-457f-a2ac-46a6258e7cd7.png)

![SD replaced](https://user-images.githubusercontent.com/82089671/229345710-405bc60d-9f15-426e-a306-078ec9bfffe3.png)

![Square - replaced](https://user-images.githubusercontent.com/82089671/229345714-f6a5955a-9a9b-4937-8155-6c46c343d776.png)

**Comment by acrylian** [comment]
Just to note that the video and it's custom/default size image for the single image is inheriting the size set on a theme's image.php. Using 720p as a default size for the video-class plugin makes sense and to disable cropping for teh sized image. 

Note that sidecar images are created in "thumbs" mode because technically the sized image is the video/audio itself. Beside that that and a "_thumb" in the cache file there is no technical difference between thumb and sized image mode.

However I would not want to override theme based settings by default. the theme sets how they are dislayed. The standard default image is not maxspace in Basic but Zenpage uses it. If you want that on your theme you need to add a switch for real images and videos (there are  methods in the image class for these checks). If that must be an option to enable.

The theme can use `object-fit` to position sidecars alternatively (which is not a replacement for properly sizes images but helps a lot).

Regardless, I will try to look into this as soon as possible.

**Comment by acrylian** [comment]
Sorry for the very late follow up. I have locally tested 16:9 ratio with maxspace posters as a default. I added an option to choose maxpace or cropping (it's not up in the repo yet). 

All works good, however since Google nowadays "demands" width/height attributes so it is requires to add those instead of the generic `width="100%" `we have currently. Then it does not fit the video player and the poster within the ratio responsively without some extra CSS:

    max-width: 100%; 
    height: auto; 
    aspect-ratio:16/9 // rather new but all not too old browser know it

For testing I have these inline. Of course if the width/height is set to non 16:9 ratios this all does not work as well and needs different CSS.

This would not apply to audio players naturally.

**Comment by acrylian** [comment]
Fixes for this with default inline css matching width/height options dynamically are up.
